Sources/Networking/v2/APIService.swift

```diff
@@ -0,0 +1,126 @@
+            return nil
+        }
+
+        try Task.checkCancellation()
+
+        // Try to decode the data
+        Logger.networking.debug("Decoding response body as \(T.self)")
+        switch T.self {
+        case is String.Type:
+            guard let resultString = String(data: data, encoding: .utf8) as? T else {
+                let error = APIRequestV2.Error.invalidDataType
+                Logger.networking.error("Error: \(error.localizedDescription)")
+                throw error
+            }
+            return resultString
+        default:
+            // Decode data
+            let decoder = JSONDecoder()
+            return try decoder.decode(T.self, from: data)
+        }
+    }
+
+    /// Fetch an API Request
+    /// - Parameter request: A configured APIRequest
+    /// - Returns: An `APIResponse`, a tuple composed by `(data: Data?, httpResponse: HTTPURLResponse)`
+    public func fetch(request: APIRequestV2) async throws -> APIService.APIResponse {
+
+        Logger.networking.debug("Fetching: \(request.debugDescription)")
+        let (data, response) = try await fetch(for: request.urlRequest)
+        Logger.networking.debug("Response: \(response.debugDescription) Data size: \(data.count) bytes")
+
+        try Task.checkCancellation()
+
+        // Check response code
+        let httpResponse = try response.asHTTPURLResponse()
+        let responseHTTPStatus = httpResponse.httpStatus
+        if responseHTTPStatus.isFailure {
+            return (data, httpResponse)
+        }
+
+        // Check requirements
+        let notModifiedIsAllowed: Bool = request.responseConstraints?.contains(.allowHTTPNotModified) ?? false
+        if responseHTTPStatus == .notModified && !notModifiedIsAllowed {
+            let error = APIRequestV2.Error.unsatisfiedRequirement(.allowHTTPNotModified)
+            Logger.networking.error("Error: \(error.localizedDescription)")
+            throw error
+        }
+        if let requirements = request.responseConstraints {
+            for requirement in requirements {
+                switch requirement {
+                case .requireETagHeader:
+                    guard httpResponse.etag != nil else {
+                        let error = APIRequestV2.Error.unsatisfiedRequirement(requirement)
+                        Logger.networking.error("Error: \(error.localizedDescription)")
+                        throw error
+                    }
+                case .requireUserAgent:
+                    guard let userAgent = httpResponse.allHeaderFields[HTTPHeaderKey.userAgent] as? String,
+                          userAgent.isEmpty == false else {
+                        let error = APIRequestV2.Error.unsatisfiedRequirement(requirement)
+                        Logger.networking.error("Error: \(error.localizedDescription)")
+                        throw error
+                    }
+                default: break
+                }
+            }
+        }
```

Comment:
Maybe we could extract this to another `validateResponse` method.

Reply:
I agree, done